File: LeetCode/57. Insert Interval.java
class Solution {
    public int[][] insert(int[][] ins, int[] newI) {
        ArrayList<int[]> list = new ArrayList<>();
        for (int i = 0; i < ins.length; i++) {
            int[] in = ins[i];
            if (in[0] > newI[1]) {
                list.add(newI);
                list.addAll(Arrays.asList(Arrays.copyOfRange(ins, i, ins.length)));
                return list.toArray(new int[0][0]);
            } else if (in[1] < newI[0]) {
                list.add(in);
            } else {
                newI[0] = Math.min(newI[0], in[0]);
                newI[1] = Math.max(newI[1], in[1]);
            }
        }
        list.add(newI);
        return list.toArray(new int[0][0]);
    }
} //TC: O(n), SC: O(n)

File: LeetCode/57. Insert Interval.kt
class Solution {
    fun insert(Is: Array<IntArray>, nI: IntArray): Array<IntArray> {
        val res = mutableListOf<IntArray>()

        Is.forEachIndexed { i, it ->
            if (it[0] > nI[1]) { //when current interval is greater than the new interval, no more overlap is possible
                res.add(nI)
                res.addAll(Is.copyOfRange(i, Is.size))
                return res.toTypedArray() //add new interval and the rest of intervals then return
            }

            if (it[1] < nI[0]) { //when overlapping has not occurred yet
                res.add(it) //add current interval
            } else { //when current interval is overlapping
                nI[0] = Math.min(it[0], nI[0])
                nI[1] = Math.max(it[1], nI[1]) //set the new interval with the lesser start and the greater end
            }
        }
        //if the new interval is the end of the array
        res.add(nI)
        return res.toTypedArray() //add it and return
    }
}
